### PR TITLE
SIMD Documentation

### PR DESCRIPTION
**Pull request by ibaned**

### CODE REVIEW

**Comment by ibaned** [comment]
This is both API and Programming Guide documentation for SIMD types. At this point it is ready for review/feedback from the Kokkos team.

File: docs/source/ProgrammingGuide/SIMD.md [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+### Conditional assignment
+
+One of the more difficult things to do with SIMD types is conditional logic. Consider the following code which is responsible for ensuring that the value `x` is not negative (we will ignore the existence of `max` functions for this discussion because it makes for a nice simple example):
+
+```c++
+double x;
+if (x < 0) x = 0;
+```
+
+We cannot naively use SIMD types in this scenario, because `x < 0` is not a boolean value, instead it is a `simnd_mask<double, Abi>` object which represents possibly multiple booleans.
```
Comment (lucbv) [inline docs/source/ProgrammingGuide/SIMD.md:186]: small typo: simnd_mask -> simd_mask

**Review by lucbv — COMMENTED** [review]
Found a small typo

File: docs/source/ProgrammingGuide/SIMD.md (hunk 1 above)
Comment (ibaned) [reply docs/source/ProgrammingGuide/SIMD.md]: Thanks @lucbv ! Just pushed a fix

**Review by dalg24 — COMMENTED** [review]
Looks good, thank you.  I only have a few minor comments.
Leaving the merge up to Francesco who is coordinating the migration to rst.

File: docs/source/API/simd/simd.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+The first template parameter `T` should be a C++ fundamental type for which the current platform supports vector intrinsics. Kokkos supports the following types for `T`:
+ - `double`
+ - `std::int32_t`
+ - `std::int64_t`
+ - `std::uint32_t`
+ - `std::uint64_t`
+
+The second template parameter `Abi` is one of the pre-defined ABI types in the namespace `Kokkos::Experimental::simd_abi`. This type determines the size of the vector and what architecture-specific intrinsics will be used. The following types are always available in that namespace:
+ - `scalar`: a fallback ABI which always has vector size of 1 and uses no special intrinsics.
+ - `native`: the "best" ABI for the architecture for which Kokkos was compiled.
```
[the remark this answers is not in the document]
Comment (ibaned) [reply docs/source/API/simd/simd.md:30]: I think it is something that in-Trilinos users want and so Kokkos overall would like to deliver, but it's unlikely that I personally will contribute it to Kokkos.

File: docs/source/API/simd/simd.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+### Width
+
+ * `static constexpr std::size_t size()`: `simd<T, Abi>::size()` is a compile-time constant of the width of the vector, i.e. the number of values of type `T` in the vector.
+
+### Constructors
+
+  * `simd()`: Default Constructor. The vector values are not initialized by this constructor.
+  * `template <class U> simd(U&&)`: Single-value constructor. The argument will be converted to `value_type` and all the vector values will be set equal to that value.
+  * `template <class G> simd(G&& gen)`: Generator constructor. The generator `gen` should be a callable type (e.g. functor) that can accept `std::integral_constant<std::size_t, i>()` as an argument and return something convertible to `value_type`. Vector lane `i` will be initialized to the value of `gen(std::integral_constant<std::size_t, i>())`.
```
[the remark this answers is not in the document]
Comment (ibaned) [reply docs/source/API/simd/simd.md:47]: Document `N4808` doesn't seem to specify that it be explicit... I can see the reasoning for it though. No preference on my end!

**Comment by ibaned** [comment]
Thanks @dalg24 ! I fixed all the things you pointed out (I think).

File: docs/source/API/simd/simd_mask.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+## Examples
+
+```c++
+#include <Kokkos_SIMD.hpp>
+#include <cstdio>
+
+int main(int argc, char* argv[]) {
+  Kokkos::initialize(argc,argv);
+  using mask_type = Kokkos::Experimental::native_simd_mask<double>;
```
Comment (dalg24) [inline docs/source/API/simd/simd_mask.md:78]: Add a scope here as well plz

**Review by dalg24 — APPROVED** [review]

File: docs/source/API/simd/simd_mask.md (hunk 4 above)
Comment (ibaned) [reply docs/source/API/simd/simd_mask.md:79]: Done! Thanks @dalg24

**Pull request merged**